cpp/src/parquet/encoding.cc

```diff
@@ -3037,12 +3083,243 @@ class RleBooleanDecoder : public DecoderImpl, virtual public BooleanDecoder {
+          while (j < common_length) {
+            if (last_value_view[j] != view[j]) {
+              break;
+            }
+            j++;
+          }
+          previous_len = src.len;
+          prefix_length_encoder_.Put({static_cast<int32_t>(j)}, 1);
+
+          last_value_view = view;
+          const auto suffix_length = static_cast<uint32_t>(src.len - j);
+          if (suffix_length == 0) {
+            suffix_encoder_.Put(&kEmpty, 1);
+            return Status::OK();
+          }
+          const uint8_t* suffix_ptr = src.ptr + j;
+          // Convert to ByteArray, so it can be passed to the suffix_encoder_.
+          const ByteArray suffix(suffix_length, suffix_ptr);
+          suffix_encoder_.Put(&suffix, 1);
+
+          return Status::OK();
+        },
+        []() { return Status::OK(); }));
+    last_value_ = last_value_view;
+  }
+
+  ::arrow::BufferBuilder sink_;
+  DeltaBitPackEncoder<Int32Type> prefix_length_encoder_;
+  DeltaLengthByteArrayEncoder<ByteArrayType> suffix_encoder_;
+  std::string last_value_;
+  const ByteArray kEmpty;
+};
+
+template <typename DType>
+void DeltaByteArrayEncoder<DType>::Put(const T* src, int num_values) {
+  throw ParquetException("Put not implemented for " + this->descr_->ToString());
+}
+
+template <>
+void DeltaByteArrayEncoder<ByteArrayType>::Put(const ByteArray* src, int num_values) {
+  if (num_values == 0) {
+    return;
+  }
+  ArrowPoolVector<int32_t> prefix_lengths(num_values,
+                                          ::arrow::stl::allocator<int32_t>(pool_));
+  std::string_view last_value_view = last_value_;
+
+  for (int i = 0; i < num_values; i++) {
+    // Convert to ByteArray, so we can pass to the suffix_encoder_.
+    const ByteArray value = src[i];
+    if (ARROW_PREDICT_FALSE(value.len >= static_cast<int32_t>(kMaxByteArraySize))) {
+      throw ParquetException("Parquet cannot store strings with size 2GB or more");
+    }
+
+    auto view = string_view{reinterpret_cast<const char*>(value.ptr),
+                            static_cast<uint32_t>(value.len)};
+    uint32_t j = 0;
+    const uint32_t common_length =
+        std::min(value.len, static_cast<uint32_t>(last_value_view.length()));
+    while (j < common_length) {
+      if (last_value_view[j] != view[j]) {
+        break;
+      }
+      j++;
+    }
+
+    last_value_view = view;
+    prefix_lengths[i] = j;
+    const auto suffix_length = static_cast<uint32_t>(value.len - j);
+
+    if (suffix_length == 0) {
+      suffix_encoder_.Put(&kEmpty, 1);
+      continue;
+    }
+    const uint8_t* suffix_ptr = value.ptr + j;
+    // Convert to ByteArray, so it can be passed to the suffix_encoder_.
+    const ByteArray suffix(suffix_length, suffix_ptr);
+    suffix_encoder_.Put(&suffix, 1);
+  }
+  prefix_length_encoder_.Put(prefix_lengths.data(), num_values);
+  last_value_ = last_value_view;
+}
+
+template <>
+void DeltaByteArrayEncoder<FLBAType>::Put(const FLBA* src, int num_values) {
```

Comment:
Wouldn't that be a bit hard to do without making `DeltaByteArrayEncoder<FLBAType>::Put` less efficient? (I'm not sure what the overhead of getting len at every iteration is)

Reply:
I'm not sure about "hard". It could be something like:
```c++
struct ByteArrayVisitor {
  const ByteArray* src;

  std::string_view operator[](int i) const {
    return std::string_view(src[i]);
  }
};
struct FLBAVisitor {
  const FLBA* src;
  int32_t len;

  std::string_view operator[](int i) const {
    return std::string_view(src[i].ptr, len);
  }
};
```

Reply:
That works indeed. Refactored the two `Put` functions to `PutInternal` (https://github.com/apache/arrow/pull/14341/commits/1d9b54a618e78d5077f1f81ea8c80f39511e4371).